### PR TITLE
Active Pathways Phase 2

### PR DESCRIPTION
**Pull request by amanpreet-github**
Removed Video dependencies for both Youtube  and custom video.

### CODE REVIEW

File: service-api/keystone.js [hunk 1]
```diff
@@ -60,7 +60,7 @@ keystone.set('routes', require('./routes'));
 keystone.set('cors allow origin', true);
 
 keystone.set('nav', {
-	'Users': ['AdminUser', 'AppUser'],
+	'Users': ['AdminUser', 'AppUser', 'Review'],
```
Comment (kushal-likhi) [inline service-api/keystone.js]: Review should be part of content.

File: service-api/routes/api/resolvers/ChallengeGetResolver.js [hunk 2]
```diff
@@ -15,7 +15,7 @@ module.exports = {
 		},
 		resolve: (parent, args, request) => (new Promise((resolve, reject) => {
 				let query = {};
-				if (!request._user && !request._user._id) return reject(new Error('Permission denied!'));
+				if (!(request._user && request._user._id)) return reject(new Error('Permission denied!'));
```
Comment (kushal-likhi) [inline service-api/routes/api/resolvers/ChallengeGetResolver.js]: Logic Changed

Comment (kushal-likhi) [reply service-api/routes/api/resolvers/ChallengeGetResolver.js]: new one looks better

File: service-api/routes/api/resolvers/UserMutationResolver.js [hunk 3]
```diff
@@ -106,52 +104,29 @@ upsertUser = (args, request, cb) => {
...
-					userChallengeState.challengeDate = moment(user.createdAt).add(i + 1, 'days').format();
-					userChallengeState.createdAt = user.createdAt;
-					userChallenges.push(userChallengeState)
-				})
-				UserChallengeState.model.insertMany(userChallenges, function (err) {
-					if (err) callback(err);
-					else callback();
+				async.series([
+					task => {
+						EmailService.sendMail(user.email, "Welcome to ActivePathways.", 'Confirmation', _user, function (err, _result) {
```
Comment (kushal-likhi) [inline service-api/routes/api/resolvers/UserMutationResolver.js]: Change Subject to `Confirm your email address`

File: service-api/routes/api/resolvers/UserMutationResolver.js [hunk 4]
```diff
@@ -106,52 +104,29 @@ upsertUser = (args, request, cb) => {
...
+							if (err) return task(err);
+							task();
+						})
+					},
+					task => {
+						EmailService.sendMail(user.email, "Welcome to ActivePathways.", 'Welcome', _user, function (err, _result) {
```
Comment (kushal-likhi) [inline service-api/routes/api/resolvers/UserMutationResolver.js]: Remove ending `Welcome`

File: service-api/routes/api/resolvers/UserProgrammeEnrollmentGetResolver.js [hunk 5]
```diff
@@ -26,6 +30,100 @@ module.exports = {
...
+						count: total,
+						status: query.status
+					});
+					reject(new Error('Invalid id!'));
+				})
+			}
+		))
+
+	},
+	UserProgrammeList: {
```
Comment (kushal-likhi) [inline service-api/routes/api/resolvers/UserProgrammeEnrollmentGetResolver.js]: Why Caps?

File: service-api/routes/api/resolvers/UserProgrammeEnrollmentGetResolver.js [hunk 6]
```diff
@@ -1,7 +1,11 @@
 const graphql = require('graphql'),
 	keystone = require('keystone'),
 	UserProgrammeEnrollment = keystone.list('UserProgrammeEnrollment').model,
-	UserProgrammeEnrollmentType = require('../types/UserProgrammeEnrollmentType');
+	async = require('async'),
+	Programme = keystone.list('Programme').model,
+	UserProgrammeEnrollmentType = require('../types/UserProgrammeEnrollmentType'),
+	EnrollmentCount = require('../types/EnrollmentCountType'),
```
Comment (kushal-likhi) [inline service-api/routes/api/resolvers/UserProgrammeEnrollmentGetResolver.js]: We do not need this, remove this file, just add a computed proprty in `ProgrammeType Node`

File: service-api/routes/api/resolvers/UserProgrammeEnrollmentGetResolver.js [hunk 7]
```diff
@@ -26,6 +30,100 @@ module.exports = {
...
+		type: new graphql.GraphQLList(ProgrammeType),
+		description: 'Fetches all programs of a user',
+		args: {
+			id: {
+				type: graphql.GraphQLString,
+				description: 'Id of user'
+			}
+		},
+		resolve: (parent, args, request) => (new Promise((resolve, reject) => {
+				let query = {}, programList = [], userEnrolledPrograms = [];
```
Comment (kushal-likhi) [inline service-api/routes/api/resolvers/UserProgrammeEnrollmentGetResolver.js]: Program , add property 'Enrolled by user' - Return -> userProgramEnrollmentType

File: service-api/routes/api/resolvers/UserProgrammeEnrollmentMutationResolver.js [hunk 8]
```diff
@@ -14,63 +18,107 @@ module.exports = {
...
+						Programme.findOne({_id: programmeId}).populate('challenges').exec((err, _programme) => {
+							if (err) return callback(err);
+
+							if (_programme) return callback(null, programme = _programme);
+
+							callback(new Error('No programme found for the given id'));
+						})
+					},
+					// Exit Previous Programs
+					callback => {
```
Comment (kushal-likhi) [inline service-api/routes/api/resolvers/UserProgrammeEnrollmentMutationResolver.js]: No need to Exit previous program

File: service-api/routes/api/resolvers/UserProgrammeEnrollmentMutationResolver.js [hunk 9]
```diff
@@ -14,63 +18,107 @@ module.exports = {
...
+					// Send Email
 					callback => {
-						if (!challenge) return callback();
-						if (args.user) challenge.user = args.user;
-						if (args.programme) challenge.programme = args.programme;
-						if (args.status) challenge.status = args.status;
-						challenge.save(function (err) {
-							if (err) callback(err);
-							else callback();
+						EmailService.sendMail(request._user.email, 'New Programme Subscribed', 'programmeEmail', programme, () => {
```
Comment (kushal-likhi) [inline service-api/routes/api/resolvers/UserProgrammeEnrollmentMutationResolver.js]: Lets test the email template

File: service-api/routes/api/types/EnrollmentCountType.js [hunk 10]
```diff
@@ -0,0 +1,14 @@
+const graphql = require('graphql');
+
+exports = module.exports = new graphql.GraphQLObjectType({
```
Comment (kushal-likhi) [inline service-api/routes/api/types/EnrollmentCountType.js]: no need for this file

File: service-api/routes/api/types/ProgrammeType.js [hunk 11]
```diff
@@ -23,6 +45,42 @@ exports = module.exports = new graphql.GraphQLObjectType({
...
+			type: graphql.GraphQLString
+		},
+		status: {
+			type: graphql.GraphQLString,
+		},
+		offeredBy: {
+			type: new graphql.GraphQLObjectType({
+				name: 'OfferedBy',
+				fields: () => ({
+					logo: {
```
Comment (kushal-likhi) [inline service-api/routes/api/types/ProgrammeType.js]: Why do we have a nested Logo Field in 'logo' ? Also should be in compatible of Cloudinary upload object.

File: service-api/routes/api/types/ReviewType.js [hunk 12]
```diff
@@ -0,0 +1,25 @@
...
+		},
+		rating: {
+			type: graphql.GraphQLInt
+		},
+		createdOn: {
+			type: graphql.GraphQLString
+		},
+		programme: {
+			type: require('./ProgrammeType'),
+			resolve: (userProgrammeEnrollment, args, request) => require('../resolvers/ProgrammeGetResolver').programmeQuery.resolve(null, {id: userProgrammeEnrollment.programme}, request),
```
Comment (kushal-likhi) [inline service-api/routes/api/types/ReviewType.js]: null check missing

File: service-api/templates/emailTemplates/Confirmation.html [hunk 13]
```diff
@@ -0,0 +1,134 @@
+<!DOCTYPE html>
+<html lang="en">
```
Comment (kushal-likhi) [inline service-api/templates/emailTemplates/Confirmation.html]: Lets get the email revieewed.

File: service-web/src/main.js [hunk 14]
```diff
@@ -9,26 +9,27 @@ import './mixins'
...
   .addPlugin(RavenVue, Vue)
   .install()
 
-Vue.use(VueYouTubeEmbed)
 Vue.use(Element)
 Vue.use(VeeValidate)
 Vue.use(VCalendar)
 Vue.use(vueScrollto)
 Vue.component(VmBackTop.name, VmBackTop)
+Vue.use(VueBlu)
```
Comment (kushal-likhi) [inline service-web/src/main.js]: what is vueBlu used for?

File: service-web/src/router/index.js [hunk 15]
```diff
@@ -21,54 +21,61 @@ const router = new Router({
...
+// router.beforeEach((to, from, next) => {
+//   // checking auth
+//   console.log('dddddd', to.name, from.name);
+//   let authCheck = LocalData.fetch.authToken();
+//   if (['dashboard', 'challengestates', 'challenge-detail'].indexOf(to.name) > -1) {
+//     if (authCheck) {
+//       next()
+//     } else {
+//       next({
+//         name: 'home-view',
```
Comment (kushal-likhi) [inline service-web/src/router/index.js]: We need this redirect functionality when not logged in.

File: service-web/src/views/ProgramPage/programPage.html [hunk 16]
```diff
@@ -1,108 +1,108 @@
...
-                <a
-                  class="f4 fw4 link dim br-pill ph4 mb4 mt4 dib white blue-dark-color pt2 pb2 lh-copy avenir pointer w-100 tc">ENROLL
+           <!-- <div class="fl w-50-m w-40-l w-100 center relative affix-tile">
+              <affix>
+                <div class="bg-white br3 pa3 pt3-m pt4-l affix-box tl absolute top-0 right-0 left-0 z-3 shadow-default">
+                  <div class="flex items-center">
+                    &lt;!&ndash; <div class="f3 f4-m f3-l fw6 mt0 flex-auto">{{isProgramJoined ? 'Enrolled in this Program' : $route.query.status === 'EXITED' ?'You have completed this program' :'Enroll to the Program'}}</div>&ndash;&gt;
+                    <div v-if="anyStatus" class="f3 f4-m f3-l fw6 mt0 flex-auto">{{$route.query.status === 'EXITED'
+                      ?'You have completed this programme.' :'You are already enrolled to the programme.'}}
+                    </div>
```
Comment (kushal-likhi) [inline service-web/src/views/ProgramPage/programPage.html]: delete unwanted code

File: service-api/routes/api/resolvers/ChallengeGetResolver.js (hunk 2 above)
Comment (amanpreet-github) [reply service-api/routes/api/resolvers/ChallengeGetResolver.js]: Thanks

File: service-api/routes/api/resolvers/UserMutationResolver.js (hunk 4 above)
Comment (amanpreet-github) [reply service-api/routes/api/resolvers/UserMutationResolver.js]: which ending welcome

File: service-api/routes/api/types/ProgrammeType.js (hunk 11 above)
Comment (amanpreet-github) [reply service-api/routes/api/types/ProgrammeType.js]: logo is an object.  We need to create sub object to get the url.

File: service-api/routes/api/types/ReviewType.js (hunk 12 above)
Comment (amanpreet-github) [reply service-api/routes/api/types/ReviewType.js]: Its done in the resolver itsself.

File: service-api/routes/api/resolvers/UserProgrammeEnrollmentMutationResolver.js (hunk 8 above)
Comment (amanpreet-github) [reply service-api/routes/api/resolvers/UserProgrammeEnrollmentMutationResolver.js]: Yooo lets remove it

File: service-api/routes/api/resolvers/UserProgrammeEnrollmentGetResolver.js (hunk 5 above)
Comment (amanpreet-github) [reply service-api/routes/api/resolvers/UserProgrammeEnrollmentGetResolver.js]: Hehe.. cool

File: service-api/routes/api/resolvers/UserMutationResolver.js (hunk 3 above)
Comment (amanpreet-github) [reply service-api/routes/api/resolvers/UserMutationResolver.js]: Yeah sure

File: service-web/src/main.js (hunk 14 above)
Comment (amanpreet-github) [reply service-web/src/main.js]: Fakru Used in making the enroll button stiky

File: service-web/src/router/index.js (hunk 15 above)
Comment (amanpreet-github) [reply service-web/src/router/index.js]: We handled it seperately...

File: service-web/src/components/challengeStates/done/done.js [hunk 17]
```diff
@@ -69,42 +62,44 @@ export default {
...
-            this.$notify({
-              title: 'Success',
-              message: 'Notes updated',
-              type: 'success'
-            });
+            self.$emit('challengeUpdated', response)
+            // self.$notify({
+            //   title: 'Success',
+            //   message: 'Notes updated',
+            //   type: 'success'
```
[the remark this answers is not in the document]
Comment (amanpreet-github) [reply service-web/src/components/challengeStates/done/done.js]: Some issue with the $notify. We need to replace it with  some other method . here some ui-element module is using. I think some dependency is not working properly

File: service-web/src/components/challengeStates/done/done.js [hunk 18]
```diff
@@ -69,42 +62,44 @@ export default {
...
-          console.log('151515', 'hit api')
           UserService.updateUserChallengeById(challengeData).then((response) => {
-            this.$emit('challengeUpdated', response)
-            this.$notify({
-              title: 'Success',
-              message: 'Notes updated',
-              type: 'success'
-            });
+            self.$emit('challengeUpdated', response)
+            // self.$notify({
```
[the remark this answers is not in the document]
Comment (amanpreet-github) [reply service-web/src/components/challengeStates/done/done.js]: See next message . Thanks

File: service-api/templates/emailTemplates/Confirmation.html (hunk 13 above)
Comment (amanpreet-github) [reply service-api/templates/emailTemplates/Confirmation.html]: Yeah reviewed.. Its fine.

File: service-api/routes/api/resolvers/UserProgrammeEnrollmentMutationResolver.js (hunk 9 above)
Comment (amanpreet-github) [reply service-api/routes/api/resolvers/UserProgrammeEnrollmentMutationResolver.js]: yeah its working.. yoo

File: service-api/routes/api/types/EnrollmentCountType.js (hunk 10 above)
Comment (amanpreet-github) [reply service-api/routes/api/types/EnrollmentCountType.js]: cool lets remove it

File: service-api/routes/api/resolvers/UserProgrammeEnrollmentGetResolver.js (hunk 7 above)
Comment (amanpreet-github) [reply service-api/routes/api/resolvers/UserProgrammeEnrollmentGetResolver.js]: ohk done ..

File: service-api/routes/api/resolvers/UserProgrammeEnrollmentGetResolver.js (hunk 6 above)
Comment (amanpreet-github) [reply service-api/routes/api/resolvers/UserProgrammeEnrollmentGetResolver.js]: cool let refactor it.. done

File: service-api/models/AppUser.js [hunk 19]
```diff
@@ -13,7 +13,8 @@ AppUser.add({
 	password: {type: Types.Password, initial: true, required: true},
 	isEnabled: {type: Types.Boolean, initial: true, default: false},
 	role: {type: Types.Select, options: ['APP_USER'], initial: true, required: true},
-	emailVerified: {type: Types.Boolean, initial: true, default: false}
+	emailVerified: {type: Types.Boolean, initial: true, default: false},
+	isNewUser: {type: Types.Boolean, initial: true, default: true}
```
Comment (kushal-likhi) [inline service-api/models/AppUser.js]: make it isOnboarded

File: service-api/routes/api/GraphQLSchema.js [hunk 20]
```diff
@@ -31,12 +31,14 @@ const ProjectNameGraphQLSchema = new graphql.GraphQLSchema({
...
 			challenge: ChallengeQueryField.challengeQuery,
 			programme: ProgrammeQueryField.programmeQuery,
+			programmeList: ProgrammeQueryField.programmeList,
 			userChallengeState: UserChallengeStateQueryField.userChallengeStateQuery,
 			userChallengeStateList: UserChallengeStateListResolver.userChallengeStateList,
 			userProgrammeEnrollment: UserProgrammeEnrollmentQueryField.userProgrammeEnrollmentQuery,
 			login: LoginQueryField.login,
 			logout: LoginQueryField.logout,
-			forgetPassword: ForgetPasswordField.forgetPassword
+			forgetPassword: ForgetPasswordField.forgetPassword,
```
Comment (kushal-likhi) [inline service-api/routes/api/GraphQLSchema.js]: unwanted comma and new line

File: service-api/routes/api/resolvers/UserMutationResolver.js [hunk 21]
```diff
@@ -62,6 +64,7 @@ upsertUser = (args, request, cb) => {
 				if (!args.id) return callback();
 				let userId = convertToObjectId(args['id']);
 				if (!userId) return callback(new Error('Invalid User ID passed!'));
+				console.log('000000------------', request._user._id.toString(), userId.toString())
```
Comment (kushal-likhi) [inline service-api/routes/api/resolvers/UserMutationResolver.js]: we can remove this log

File: service-api/routes/api/types/ProgrammeType.js (hunk 11 above)
Comment (kushal-likhi) [reply service-api/routes/api/types/ProgrammeType.js]: looks correct,, sorry man : )

File: service-web/src/views/ProgramsCatalogue/programs-catalogue.js [hunk 22]
```diff
@@ -8,10 +10,47 @@ export default {
...
+    user() {
+      return this.$store.state.auth.user
+    }
+  },
+  methods: {
+    fetchProgrammeList() {
+      const self = this;
+      ProgramService.getProgrammeList(100, 0)
+        .then(programmeList => {
+          this.programs = programmeList.sort((programmeA, programmeB) => {
```
Comment (kushal-likhi) [inline service-web/src/views/ProgramsCatalogue/programs-catalogue.js]: we can make a util method `sortUserProgrammes(<prgs>): <prgs sorted>`

**Pull request merged**